src/psyclone/domain/lfric/kernel/lfric_kernel_metadata.py

```diff
@@ -0,0 +1,506 @@
+
+    :param meta_args: a list of 'meta_arg' objects which capture the \
+        metadata values of the kernel arguments.
+    :type meta_args: Optional[List[:py:class:`ScalarArg` | \
+        :py:class:`FieldArg` | :py:class:`OperatorArg`]]
+    :param meta_funcs: a list of 'meta_arg' objects which capture the \
+        metadata values of the kernel arguments.
+    :type meta_funcs: Optional[List[:py:class:`GridArg` | \
+        :py:class:`FieldArg` | :py:class:`ScalarArg`]]
+    :param meta_reference_element: a list of 'meta_arg' objects which \
+        capture the metadata values of the kernel arguments.
+    :type meta_reference_element: Optional[List[:py:class:`GridArg` | \
+        :py:class:`FieldArg` | :py:class:`ScalarArg`]]
+    :param meta_mesh: a list of 'meta_arg' objects which capture the \
+        metadata values of the kernel arguments.
+    :type meta_mesh: Optional[List[:py:class:`ScalarArg` | \
+        :py:class:`FieldArg` | :py:class:`OperatorArg`]]
+    :param shape: quadrature.
+    :type shape: Optional[str]
+    :param operates_on: the name of the quantity that this kernel is
+        intended to iterate over.
+    :type operates_on: Optional[str]
+    :param procedure_name: the name of the kernel procedure to call.
+    :type procedure_name: Optional[str]
+    :param name: the name of the symbol to use for the metadata in \
+        language-level PSyIR.
+    :type name: Optional[str]
+
+    raises TypeError: if meta args is not a list of argument objects.
+
+    '''
+    VALID_NAME = re.compile(r'[a-zA-Z_][\w]*')
+
+    def __init__(self, operates_on=None, gh_shape=None, meta_args=None,
+                 meta_funcs=None, meta_reference_element=None,
+                 meta_mesh=None, procedure_name=None, name=None):
+        # Validate values using setters if they are not None
+        self._operates_on = None
+        if operates_on is not None:
+            self.operates_on = operates_on
+        self._gh_shape = None
+        if gh_shape is not None:
+            self.gh_shape = gh_shape
+            # TBD
+        if meta_args is None:
+            self._meta_args = []
+        else:
+            if not isinstance(meta_args, list):
+                raise TypeError(f"meta_args should be a list but found "
+                                f"{type(meta_args).__name__}.")
+            for entry in meta_args:
+                if not isinstance(entry, CommonArg):
+                    raise TypeError(
+                        f"meta_args should be a list of argument objects, "
+                        f"but found {type(entry).__name__}.")
+            self._meta_args = meta_args
+        if meta_funcs is None:
+            self._meta_funcs = []
+            # TBD
+        if meta_reference_element is None:
+            self._meta_reference_element = []
+            # TBD
+        if meta_mesh is None:
+            self._meta_mesh = []
+            # TBD
+
+        self._procedure_name = None
```

Comment:
You and your fancy ways.

Reply:
already done

Reply:
No, as I realised we can't do this. If there is a procedure name provided we validate it with the setter, otherwise we simply set the internal value to None without validation. I added comments to say this.

Reply:
Ah, yes.